cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/io/cdap/cdap/etl/common/OAuthAccessTokenMacroEvaluatorTest.java

```diff
@@ -0,0 +1,141 @@
+import io.cdap.cdap.common.service.ServiceDiscoverable;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.security.auth.context.AuthenticationTestContext;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HttpResponder;
+import io.cdap.http.NettyHttpService;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.InMemoryDiscoveryService;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.Map;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ *
+ */
+public class OAuthAccessTokenMacroEvaluatorTest {
+
+  private static final Gson GSON = new Gson();
+  private static final String PROVIDER = "test";
+  private static final String CREDENTIAL_ID = "testcredential";
+
+  private static NettyHttpService httpService;
+  private static ServiceDiscoverer serviceDiscoverer;
+
+  @BeforeClass
+  public static void init() throws Exception {
+    httpService = NettyHttpService.builder("OAuthTest")
+      .setHttpHandlers(new OAuthHandler(
+        ImmutableMap.of(
+          PROVIDER, ImmutableMap.of(CREDENTIAL_ID, new OAuthInfo("foobar", "bearer"))
+        )))
+      .build();
+
+    httpService.start();
+
+    InMemoryDiscoveryService discoveryService = new InMemoryDiscoveryService();
+    String discoveryName = ServiceDiscoverable.getName(NamespaceId.SYSTEM.getNamespace(),
+                                                       Constants.PIPELINEID, ProgramType.SERVICE,
+                                                       Constants.STUDIO_SERVICE_NAME);
+    discoveryService.register(new Discoverable(discoveryName, httpService.getBindAddress()));
+    RemoteClientFactory remoteClientFactory = new RemoteClientFactory(
+      discoveryService, new DefaultInternalAuthenticator(new AuthenticationTestContext()));
+    serviceDiscoverer = new AbstractServiceDiscoverer(NamespaceId.DEFAULT.app("testapp").spark("testspark")) {
+      @Override
+      protected RemoteClientFactory getRemoteClientFactory() {
+        return remoteClientFactory;
+      }
+    };
+  }
+
+  @AfterClass
+  public static void finish() throws Exception {
+    httpService.stop();
+  }
+
+  @Test
+  public void testOAuthAccessTokenMacro() {
+    MacroEvaluator macroEvaluator = new OAuthAccessTokenMacroEvaluator(serviceDiscoverer);
+    String accessToken = macroEvaluator.evaluate(OAuthAccessTokenMacroEvaluator.FUNCTION_NAME,
+                                                                PROVIDER, CREDENTIAL_ID);
+    // assert contain all properties
+    Assert.assertEquals("foobar", accessToken);
+  }
+
+  /**
+   * A http handler to provide the OAuth endpoint.
+   */
+  public static final class OAuthHandler extends AbstractHttpHandler {
```

Comment:
This seems to be duplicated from the OAuthMacroEvaluatorTest. Can we extract this out into its own class, i.e. `MockOAuthHandler`?

Reply:
done